modules/core/src/main/scala/fs2/kafka/TransactionalKafkaProducer.scala

```diff
@@ -42,90 +42,124 @@ abstract class TransactionalKafkaProducer[F[_], K, V] {
 private[kafka] object TransactionalKafkaProducer {
-  def resource[F[_], K, V](
-    settings: ProducerSettings[F, K, V]
-  )(
+
+  /**
+    * Support class capturing core functionality required to implement
+    * a [[TransactionalKafkaProducer]], allowing users to decide how transactional
+    * IDs are computed as a follow-up step to registering producer settings.
+    */
+  class Builder[F[_], K, V] private[kafka] (settings: ProducerSettings[F, K, V])(
     implicit F: ConcurrentEffect[F],
     context: ContextShift[F]
-  ): Resource[F, TransactionalKafkaProducer[F, K, V]] =
-    Resource.liftF(settings.keySerializer).flatMap { keySerializer =>
-      Resource.liftF(settings.valueSerializer).flatMap { valueSerializer =>
-        WithProducer(settings)
-          .evalTap { withProducer =>
-            withProducer { producer =>
-              F.delay(producer.initTransactions())
-            }
-          }
-          .map { withProducer =>
-            new TransactionalKafkaProducer[F, K, V] {
-              override def produce[G[+_], P](
-                records: TransactionalProducerRecords[F, G, K, V, P]
-              ): F[ProducerResult[Chunk, K, V, P]] =
-                produceTransaction(records)
-                  .map(ProducerResult(_, records.passthrough))
+  ) {
 
-              private[this] def produceTransaction[G[+_], P](
-                records: TransactionalProducerRecords[F, G, K, V, P]
-              ): F[Chunk[(ProducerRecord[K, V], RecordMetadata)]] = {
-                if (records.records.isEmpty) F.pure(Chunk.empty)
-                else {
-                  val batch =
-                    CommittableOffsetBatch.fromFoldableMap(records.records)(_.offset)
+    type Records = CommittableProducerRecords[F, K, V]
+    type RecordsChunk = Chunk[Records]
+    type RecordGrouper = RecordsChunk => F[Chunk[(KafkaByteProducer, RecordsChunk)]]
 
-                  val consumerGroupId =
-                    if (batch.consumerGroupIdsMissing || batch.consumerGroupIds.size != 1)
-                      F.raiseError(ConsumerGroupException(batch.consumerGroupIds))
-                    else F.pure(batch.consumerGroupIds.head)
+    private[this] def resource(
+      buildGrouper: Blocker => Resource[F, RecordGrouper]
+    ): Resource[F, TransactionalKafkaProducer[F, K, V]] =
+      Resource.liftF(settings.keySerializer).flatMap { keySerializer =>
+        Resource.liftF(settings.valueSerializer).flatMap { valueSerializer =>
+          settings.blocker.fold(Blockers.producer)(Resource.pure[F, Blocker]).flatMap { blocker =>
+            buildGrouper(blocker).map { grouper =>
+              new TransactionalKafkaProducer[F, K, V] {
+                override def produce[P](
+                  records: TransactionalProducerRecords[F, K, V, P]
+                ): F[ProducerResult[Chunk, K, V, P]] =
+                  grouper(records.records)
+                    .flatMap { recordGroups =>
+                      recordGroups.flatTraverse {
```

Comment:
Should these run in parallel rather than sequentially?

Reply:
:+1: I think they should, we just need the extra type parameter.